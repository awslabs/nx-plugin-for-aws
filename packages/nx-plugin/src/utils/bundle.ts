/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import { TargetConfiguration } from '@nx/devkit';

export interface CreatePythonBundleTargetOptions {
  /**
   * Directory of the python project from the monorepo root
   */
  projectDir: string;

  /**
   * Python package name
   */
  packageName: string;
}

/**
 * Create a target for bundling a python project
 */
export const createPythonBundleTarget = ({
  projectDir,
  packageName,
}: CreatePythonBundleTargetOptions): TargetConfiguration => {
  return {
    cache: true,
    executor: 'nx:run-commands',
    outputs: [`{workspaceRoot}/dist/${projectDir}/bundle`],
    options: {
      commands: [
        `uv export --frozen --no-dev --no-editable --project ${projectDir} --package ${packageName} -o dist/${projectDir}/bundle/requirements.txt`,
        `uv pip install -n --no-deps --no-installer-metadata --no-compile-bytecode --python-platform x86_64-manylinux2014 --target dist/${projectDir}/bundle -r dist/${projectDir}/bundle/requirements.txt`,
      ],
      parallel: false,
    },
  };
};
