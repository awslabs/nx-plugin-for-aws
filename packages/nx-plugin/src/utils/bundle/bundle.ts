/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  addDependenciesToPackageJson,
  generateFiles,
  joinPathFragments,
  OverwriteStrategy,
  ProjectConfiguration,
  TargetConfiguration,
  Tree,
} from '@nx/devkit';
import { withVersions } from '../versions';
import { query, replace } from '../ast';
import {
  ArrayLiteralExpression,
  factory,
  ObjectLiteralExpression,
  PropertyAssignment,
  SyntaxKind,
} from 'typescript';
import { StringLiteral } from '@phenomnomnominal/tsquery';
import { getRelativePathToRoot } from '../paths';
import { addDependencyToTargetIfNotPresent } from '../nx';

export interface AddPythonBundleTargetOptions {
  /**
   * Python platform
   * @default x86_64-manylinux2014
   */
  pythonPlatform?: 'x86_64-manylinux2014' | 'aarch64-manylinux2014';
}

interface CreatePythonBundleTargetOptions
  extends Required<AddPythonBundleTargetOptions> {
  /**
   * Directory of the python project from the monorepo root
   */
  projectDir: string;

  /**
   * Python package name
   */
  packageName: string;

  /**
   * Name of the bundle target, used for the outdir for the bundle
   */
  bundleTargetName: string;
}

/**
 * Create a target for bundling a python project
 */
const createPythonBundleTarget = ({
  projectDir,
  packageName,
  pythonPlatform,
  bundleTargetName,
}: CreatePythonBundleTargetOptions): TargetConfiguration => {
  return {
    cache: true,
    executor: 'nx:run-commands',
    outputs: [`{workspaceRoot}/dist/${projectDir}/${bundleTargetName}`],
    options: {
      commands: [
        `uv export --frozen --no-dev --no-editable --project ${projectDir} --package ${packageName} -o dist/${projectDir}/${bundleTargetName}/requirements.txt`,
        `uv pip install -n --no-deps --no-installer-metadata --no-compile-bytecode --python-platform ${pythonPlatform} --target dist/${projectDir}/${bundleTargetName} -r dist/${projectDir}/${bundleTargetName}/requirements.txt`,
      ],
      parallel: false,
    },
  };
};

/**
 * Adds a bundle target to the given project if it does not exist, and updates the build target to depend on it
 */
export const addPythonBundleTarget = (
  project: ProjectConfiguration,
  opts?: AddPythonBundleTargetOptions,
) => {
  if (!project.targets) {
    project.targets = {};
  }

  const pythonPlatform = opts?.pythonPlatform ?? 'x86_64-manylinux2014';
  const bundleTargetName =
    pythonPlatform === 'aarch64-manylinux2014' ? 'bundle-arm' : 'bundle-x86';

  if (!project.targets?.[bundleTargetName]) {
    project.targets[bundleTargetName] = {
      ...createPythonBundleTarget({
        projectDir: project.root,
        packageName: project.name,
        pythonPlatform,
        bundleTargetName,
      }),
      dependsOn: ['compile'],
    };
  }

  // Add a "bundle" target which depends on either bundle-arm or bundle-x86 (or both)
  addDependencyToTargetIfNotPresent(project, 'bundle', bundleTargetName);
  addDependencyToTargetIfNotPresent(project, 'build', 'bundle');

  return {
    bundleTargetName,
    bundleOutputDir: joinPathFragments('dist', project.root, bundleTargetName),
  };
};

export interface AddTypeScriptBundleTargetOptions {
  /**
   * Path to the target file relative to the project dir
   */
  targetFilePath: string;

  /**
   * Sub directory to write bundled index.js file to (if any)
   * Outputs to dist/{projectRoot}/bundle/{bundleOutputDir}/index.js
   */
  bundleOutputDir?: string;

  /**
   * Modules to omit from the bundle and treat as external
   */
  external?: (string | RegExp)[];

  /**
   * Target platform for the bundle
   * @default 'node'
   */
  platform?: 'node' | 'browser' | 'neutral';
}

/**
 * Add a TypeScript bundle target using rolldown
 */
export const addTypeScriptBundleTarget = (
  tree: Tree,
  project: ProjectConfiguration,
  opts: AddTypeScriptBundleTargetOptions,
) => {
  project.targets ??= {};

  // Check if AWS SDK is external
  const awsSdkRegex = /@aws-sdk\/.*/;
  const isAwsSdkExternal = opts.external?.some((ext) =>
    ext instanceof RegExp
      ? ext.source === awsSdkRegex.source
      : ext === awsSdkRegex.source,
  );

  // Generate empty rolldown config if it doesn't exist
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files', 'ts'),
    project.root,
    {},
    { overwriteStrategy: OverwriteStrategy.KeepExisting },
  );

  // Add the bundle target
  if (!project.targets.bundle) {
    project.targets.bundle = {
      cache: true,
      outputs: [`{workspaceRoot}/dist/{projectRoot}/bundle`],
      executor: 'nx:run-commands',
      options: {
        command: 'rolldown -c rolldown.config.ts',
        cwd: '{projectRoot}',
      },
      dependsOn: ['compile'],
    };
  }

  // Add bundle to the build target
  addDependencyToTargetIfNotPresent(project, 'build', 'bundle');

  const rolldownConfigPath = joinPathFragments(
    project.root,
    'rolldown.config.ts',
  );

  const rolldownConfigArraySelector =
    'CallExpression:has(Identifier[name="defineConfig"]) > ArrayLiteralExpression';

  // Add disableTreeShake function if AWS SDK is not external and it doesn't exist
  // This is due to an issue with rolldown bundling for the SDK, and can likely be removed when resolved
  // https://github.com/rolldown/rolldown/issues/6513
  if (!isAwsSdkExternal) {
    const disableTreeShakeSelector =
      'VariableDeclaration:has(Identifier[name="disableTreeShake"])';
    if (
      query(tree, rolldownConfigPath, disableTreeShakeSelector).length === 0
    ) {
      const content = tree.read(rolldownConfigPath, 'utf-8');
      const disableTreeShakeFunction = `// Disables tree-shaking for the given module patterns
const disableTreeShake = (patterns: RegExp[]) => ({
  name: 'disable-treeshake',
  transform: (code, id) => {
    if (patterns.some(p => p.test(id))) {
      return { code, map: null, moduleSideEffects: 'no-treeshake' };
    }
    return null;
  },
});

`;
      const defineConfigIndex = content.indexOf('export default defineConfig');
      if (defineConfigIndex !== -1) {
        const newContent =
          content.slice(0, defineConfigIndex) +
          disableTreeShakeFunction +
          content.slice(defineConfigIndex);
        tree.write(rolldownConfigPath, newContent);
      }
    }
  }

  // Check whether we already have a config entry with input set to targetFilePath
  if (
    query(
      tree,
      rolldownConfigPath,
      `${rolldownConfigArraySelector} PropertyAssignment:has(Identifier[name="input"]):has(StringLiteral[value="${opts.targetFilePath}"])`,
    ).length === 0
  ) {
    // We don't have one, so append it
    replace(
      tree,
      rolldownConfigPath,
      rolldownConfigArraySelector,
      (node: ArrayLiteralExpression) => {
        return factory.createArrayLiteralExpression([
          ...node.elements,
          factory.createObjectLiteralExpression(
            [
              factory.createPropertyAssignment(
                factory.createIdentifier('input'),
                factory.createStringLiteral(opts.targetFilePath, true),
              ),
              factory.createPropertyAssignment(
                factory.createIdentifier('output'),
                factory.createObjectLiteralExpression(
                  [
                    factory.createPropertyAssignment(
                      factory.createIdentifier('file'),
                      factory.createStringLiteral(
                        joinPathFragments(
                          getRelativePathToRoot(tree, project.name),
                          'dist',
                          project.root,
                          'bundle',
                          opts.bundleOutputDir ?? '.',
                          'index.js',
                        ),
                        true,
                      ),
                    ),
                    factory.createPropertyAssignment(
                      factory.createIdentifier('format'),
                      factory.createStringLiteral('cjs', true),
                    ),
                    factory.createPropertyAssignment(
                      factory.createIdentifier('inlineDynamicImports'),
                      factory.createTrue(),
                    ),
                  ],
                  true,
                ),
              ),
              factory.createPropertyAssignment(
                factory.createIdentifier('platform'),
                factory.createStringLiteral(opts.platform ?? 'node', true),
              ),
              ...(opts.external
                ? [
                    factory.createPropertyAssignment(
                      factory.createIdentifier('external'),
                      factory.createArrayLiteralExpression(
                        opts.external.map((ext) =>
                          typeof ext === 'string'
                            ? factory.createStringLiteral(ext, true)
                            : factory.createRegularExpressionLiteral(
                                `/${ext.source}/`,
                              ),
                        ),
                      ),
                    ),
                  ]
                : []),
              ...(!isAwsSdkExternal
                ? [
                    factory.createPropertyAssignment(
                      factory.createIdentifier('plugins'),
                      factory.createArrayLiteralExpression([
                        factory.createCallExpression(
                          factory.createIdentifier('disableTreeShake'),
                          undefined,
                          [
                            factory.createArrayLiteralExpression([
                              factory.createRegularExpressionLiteral(
                                `/${awsSdkRegex.source}/`,
                              ),
                            ]),
                          ],
                        ),
                      ]),
                    ),
                  ]
                : []),
            ],
            true,
          ),
        ]);
      },
    );
  }

  addDependenciesToPackageJson(tree, {}, withVersions(['rolldown']));
};
