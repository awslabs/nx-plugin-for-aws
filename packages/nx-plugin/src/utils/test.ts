/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { vi, expect } from 'vitest';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { joinPathFragments, Tree } from '@nx/devkit';
import solutionSetup from '@nx/js/src/utils/typescript/ts-solution-setup';

export const createTreeUsingTsSolutionSetup = (): Tree => {
  vi.spyOn(solutionSetup, 'isUsingTsSolutionSetup').mockImplementation(
    () => true,
  );

  const tree = createTreeWithEmptyWorkspace();

  tree.write('pnpm-workspace.yaml', `packages:\n  - packages/*`);

  tree.write('tsconfig.json', '{}');
  return tree;
};

/**
 * Snapshot all files within a directory in the given tree
 */
export const snapshotTreeDir = (tree: Tree, dir: string) => {
  if (tree.isFile(dir)) {
    expect(tree.read(dir, 'utf-8')).toMatchSnapshot(dir);
  } else {
    tree
      .children(dir)
      .forEach((subDir) =>
        snapshotTreeDir(tree, joinPathFragments(dir, subDir)),
      );
  }
};
