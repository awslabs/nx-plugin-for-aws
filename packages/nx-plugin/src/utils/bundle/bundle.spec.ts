/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  Tree,
  ProjectConfiguration,
  readJson,
  addProjectConfiguration,
} from '@nx/devkit';
import { createTreeUsingTsSolutionSetup } from '../test';
import { addTypeScriptBundleTarget, addPythonBundleTarget } from './bundle';

describe('bundle utilities', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
  });

  describe('addTypeScriptBundleTarget', () => {
    let project: ProjectConfiguration;

    beforeEach(() => {
      project = {
        name: 'test-project',
        root: 'apps/test-project',
        sourceRoot: 'apps/test-project/src',
        targets: {},
      };
      addProjectConfiguration(tree, project.name, project);
    });

    it('should add bundle target to project configuration', () => {
      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
      });

      expect(project.targets?.bundle).toBeDefined();
      expect(project.targets?.bundle?.cache).toBe(true);
      expect(project.targets?.bundle?.executor).toBe('nx:run-commands');
      expect(project.targets?.bundle?.outputs).toEqual([
        '{workspaceRoot}/dist/{projectRoot}/bundle',
      ]);
      expect(project.targets?.bundle?.options?.command).toBe(
        'rolldown -c rolldown.config.ts',
      );
      expect(project.targets?.bundle?.options?.cwd).toBe('{projectRoot}');
    });

    it('should not overwrite existing bundle target', () => {
      const existingBundleTarget = {
        cache: false,
        executor: 'custom:executor',
        options: { customOption: 'value' },
      };
      project.targets = { bundle: existingBundleTarget };

      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
      });

      expect(project.targets.bundle).toEqual(existingBundleTarget);
    });

    it('should generate rolldown config file', () => {
      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
      });

      expect(tree.exists('apps/test-project/rolldown.config.ts')).toBeTruthy();
    });

    it('should not overwrite existing rolldown config file', () => {
      const existingConfig = `
import { defineConfig } from 'rolldown';

export default defineConfig([
  {
    input: 'existing/file.ts',
    output: {
      file: 'existing/output.js',
      format: 'cjs',
    },
  },
]);
`;
      tree.write('apps/test-project/rolldown.config.ts', existingConfig);

      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
      });

      const configContent = tree.read(
        'apps/test-project/rolldown.config.ts',
        'utf-8',
      );
      expect(configContent).toContain('existing/file.ts');
      expect(configContent).toContain('src/index.ts');
    });

    it('should add build target dependency on bundle', () => {
      project.targets = {
        build: {
          executor: 'nx:run-commands',
          dependsOn: ['compile'],
        },
      };

      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
      });

      expect(project.targets.build.dependsOn).toContain('bundle');
      expect(project.targets.build.dependsOn).toContain('compile');
    });

    it('should not duplicate bundle dependency in build target', () => {
      project.targets = {
        build: {
          executor: 'nx:run-commands',
          dependsOn: ['compile', 'bundle'],
        },
      };

      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
      });

      const bundleDependencies = project.targets.build.dependsOn?.filter(
        (dep) => dep === 'bundle',
      );
      expect(bundleDependencies).toHaveLength(1);
    });

    it('should handle project without existing targets', () => {
      const projectWithoutTargets: ProjectConfiguration = {
        name: 'test-project',
        root: 'apps/test-project',
        sourceRoot: 'apps/test-project/src',
      };

      addTypeScriptBundleTarget(tree, projectWithoutTargets, {
        targetFilePath: 'src/index.ts',
      });

      expect(projectWithoutTargets.targets).toBeDefined();
      expect(projectWithoutTargets.targets?.bundle).toBeDefined();
    });

    it('should configure output path with bundleOutputDir', () => {
      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
        bundleOutputDir: 'lambda',
      });

      const configContent = tree.read(
        'apps/test-project/rolldown.config.ts',
        'utf-8',
      );
      expect(configContent).toContain(
        'dist/apps/test-project/bundle/lambda/index.js',
      );
    });

    it('should configure output path without bundleOutputDir', () => {
      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
      });

      const configContent = tree.read(
        'apps/test-project/rolldown.config.ts',
        'utf-8',
      );
      expect(configContent).toContain('dist/apps/test-project/bundle/index.js');
    });

    it('should add rolldown dependency to package.json', () => {
      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
      });

      const packageJson = readJson(tree, 'package.json');
      expect(packageJson.devDependencies).toHaveProperty('rolldown');
    });

    it('should not add duplicate config entry for same targetFilePath', () => {
      // First call
      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
      });

      // Second call with same targetFilePath
      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
      });

      const configContent = tree.read(
        'apps/test-project/rolldown.config.ts',
        'utf-8',
      );

      // Count occurrences of the input configuration
      const inputOccurrences = (
        configContent.match(/input:\s*['"]src\/index\.ts['"]/g) || []
      ).length;
      expect(inputOccurrences).toBe(1);
    });

    it('should add multiple config entries for different targetFilePaths', () => {
      // First call
      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
      });

      // Second call with different targetFilePath
      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/handler.ts',
        bundleOutputDir: 'handler',
      });

      const configContent = tree.read(
        'apps/test-project/rolldown.config.ts',
        'utf-8',
      );

      expect(configContent).toContain('src/index.ts');
      expect(configContent).toContain('src/handler.ts');
      expect(configContent).toContain('bundle/index.js');
      expect(configContent).toContain('bundle/handler/index.js');
    });

    it('should configure rolldown with correct format and options', () => {
      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
      });

      const configContent = tree.read(
        'apps/test-project/rolldown.config.ts',
        'utf-8',
      );

      expect(configContent).toContain("format: 'cjs'");
      expect(configContent).toContain('inlineDynamicImports: true');
    });

    it('should work with nested project structure', () => {
      const nestedProject: ProjectConfiguration = {
        name: 'nested-project',
        root: 'libs/nested/project',
        sourceRoot: 'libs/nested/project/src',
        targets: {},
      };
      addProjectConfiguration(tree, nestedProject.name, nestedProject);

      addTypeScriptBundleTarget(tree, nestedProject, {
        targetFilePath: 'src/lib.ts',
      });

      expect(
        tree.exists('libs/nested/project/rolldown.config.ts'),
      ).toBeTruthy();

      const configContent = tree.read(
        'libs/nested/project/rolldown.config.ts',
        'utf-8',
      );
      expect(configContent).toContain(
        'dist/libs/nested/project/bundle/index.js',
      );
    });

    it('should configure external dependencies with strings', () => {
      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
        external: ['@aws-sdk/*', 'lodash'],
      });

      const configContent = tree.read(
        'apps/test-project/rolldown.config.ts',
        'utf-8',
      );

      expect(configContent).toContain("external: ['@aws-sdk/*', 'lodash']");
    });

    it('should configure external dependencies with regex patterns', () => {
      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
        external: [/@aws-sdk\/.*/, /^lodash/],
      });

      const configContent = tree.read(
        'apps/test-project/rolldown.config.ts',
        'utf-8',
      );

      expect(configContent).toContain('external: [/@aws-sdk\\/.*/, /^lodash/]');
    });

    it('should configure external dependencies with mixed strings and regex', () => {
      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
        external: ['@aws-sdk/*', /@types\/.*/, 'react'],
      });

      const configContent = tree.read(
        'apps/test-project/rolldown.config.ts',
        'utf-8',
      );

      expect(configContent).toContain(
        "external: ['@aws-sdk/*', /@types\\/.*/, 'react']",
      );
    });

    it('should not include external property when not provided', () => {
      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
      });

      const configContent = tree.read(
        'apps/test-project/rolldown.config.ts',
        'utf-8',
      );

      expect(configContent).not.toContain('external:');
    });

    it('should create build target if not present', () => {
      // Project has no build target initially
      expect(project.targets?.build).toBeUndefined();

      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
      });

      expect(project.targets?.build).toBeDefined();
      expect(project.targets?.build?.dependsOn).toContain('bundle');
    });

    it('should create build target with bundle dependency even when no other dependencies exist', () => {
      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
      });

      expect(project.targets?.build).toBeDefined();
      expect(project.targets?.build?.dependsOn).toEqual(['bundle']);
    });

    it('should handle external dependencies in multiple config entries', () => {
      // First call with external dependencies
      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/index.ts',
        external: ['@aws-sdk/*'],
      });

      // Second call with different external dependencies
      addTypeScriptBundleTarget(tree, project, {
        targetFilePath: 'src/handler.ts',
        external: ['lodash', /@types\/.*/],
        bundleOutputDir: 'handler',
      });

      const configContent = tree.read(
        'apps/test-project/rolldown.config.ts',
        'utf-8',
      );

      expect(configContent).toContain("external: ['@aws-sdk/*']");
      expect(configContent).toContain("external: ['lodash', /@types\\/.*/]");
      expect(configContent).toContain('src/index.ts');
      expect(configContent).toContain('src/handler.ts');
    });
  });

  describe('addPythonBundleTarget', () => {
    let project: ProjectConfiguration;

    beforeEach(() => {
      project = {
        name: 'test-python-project',
        root: 'apps/test-python-project',
        sourceRoot: 'apps/test-python-project/src',
        targets: {},
      };
      addProjectConfiguration(tree, project.name, project);
    });

    it('should add python bundle target with default platform', () => {
      addPythonBundleTarget(project);

      expect(project.targets?.bundle).toBeDefined();
      expect(project.targets?.bundle?.cache).toBe(true);
      expect(project.targets?.bundle?.executor).toBe('nx:run-commands');
      expect(project.targets?.bundle?.dependsOn).toContain('compile');

      const commands = project.targets?.bundle?.options?.commands;
      expect(commands[1]).toContain('--python-platform x86_64-manylinux2014');
    });

    it('should add python bundle target with custom platform', () => {
      addPythonBundleTarget(project, {
        pythonPlatform: 'aarch64-manylinux2014',
      });

      const commands = project.targets?.bundle?.options?.commands;
      expect(commands[1]).toContain('--python-platform aarch64-manylinux2014');
    });

    it('should not overwrite existing bundle target', () => {
      const existingBundleTarget = {
        cache: false,
        executor: 'custom:executor',
      };
      project.targets = { bundle: existingBundleTarget };

      addPythonBundleTarget(project);

      expect(project.targets.bundle).toEqual(existingBundleTarget);
    });

    it('should add bundle dependency to build target', () => {
      project.targets = {
        build: {
          executor: 'nx:run-commands',
          dependsOn: ['test'],
        },
      };

      addPythonBundleTarget(project);

      expect(project.targets.build.dependsOn).toContain('bundle');
      expect(project.targets.build.dependsOn).toContain('test');
    });

    it('should handle project without existing targets', () => {
      const projectWithoutTargets: ProjectConfiguration = {
        name: 'test-python-project',
        root: 'apps/test-python-project',
      };

      addPythonBundleTarget(projectWithoutTargets);

      expect(projectWithoutTargets.targets).toBeDefined();
      expect(projectWithoutTargets.targets?.bundle).toBeDefined();
    });
  });
});
