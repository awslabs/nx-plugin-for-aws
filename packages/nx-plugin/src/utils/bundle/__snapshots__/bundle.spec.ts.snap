// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bundle utilities > addTypeScriptBundleTarget > should add disableTreeShake function and plugins when AWS SDK is not external 1`] = `
"import { defineConfig } from 'rolldown';

// Disables tree-shaking for the given module patterns
const disableTreeShake = (patterns: RegExp[]) => ({
  name: 'disable-treeshake',
  transform: (code, id) => {
    if (patterns.some(p => p.test(id))) {
      return { code, map: null, moduleSideEffects: 'no-treeshake' };
    }
    return null;
  },
});

export default defineConfig([{
        tsconfig: 'tsconfig.lib.json',
        input: 'src/index.ts',
        output: {
            file: '../../dist/apps/test-project/bundle/index.js',
            format: 'cjs',
            inlineDynamicImports: true
        },
        platform: 'node',
        plugins: [disableTreeShake([/@aws-sdk\\/.*/])]
    }]);
"
`;

exports[`bundle utilities > addTypeScriptBundleTarget > should not add disableTreeShake or plugins when AWS SDK is external as regex 1`] = `
"import { defineConfig } from 'rolldown';

export default defineConfig([{
        tsconfig: 'tsconfig.lib.json',
        input: 'src/index.ts',
        output: {
            file: '../../dist/apps/test-project/bundle/index.js',
            format: 'cjs',
            inlineDynamicImports: true
        },
        platform: 'node',
        external: [/@aws-sdk\\/.*/]
    }]);
"
`;
